algebird-core/src/main/scala/com/twitter/algebird/maps/increment.scala

```diff
@@ -0,0 +1,159 @@
+    /** The monoid that defines what it means to increment a value */
+    def valueMonoid: Monoid[V]
+
+    /** Increment the value at a key, by another value */
+    private[increment] final def incr(di: DataMap[K, V]) = blacken(inc(di))
+
+    private[tree] def inc(di: DataMap[K, V]): Node[K]
+  }
+
+  /** Leaf R/B node supporting increment */
+  trait LNodeInc[K, V] extends NodeInc[K, V] with LNodeMap[K, V] {
+    final def inc(di: DataMap[K, V]) = {
+      val d = new DataMap[K, V] {
+        val key = di.key
+        val value = valueMonoid.plus(valueMonoid.zero, di.value)
+      }
+      rNode(d, this, this)
+    }
+  }
+
+  /** Internal R/B node supporting increment */
+  trait INodeInc[K, V] extends NodeInc[K, V] with INodeMap[K, V] {
+    def lsub: NodeInc[K, V]
+    def rsub: NodeInc[K, V]
+
+    final def inc(di: DataMap[K, V]) =
+      if (color == R) {
+        if (keyOrdering.lt(di.key, data.key)) rNode(data, lsub.inc(di), rsub)
+        else if (keyOrdering.gt(di.key, data.key)) rNode(data, lsub, rsub.inc(di))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          rNode(d, lsub, rsub)
+        }
+      } else {
+        if (keyOrdering.lt(di.key, data.key)) balance(bNode(data, lsub.inc(di), rsub))
+        else if (keyOrdering.gt(di.key, data.key)) balance(bNode(data, lsub, rsub.inc(di)))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          bNode(d, lsub, rsub)
+        }
+      }
+  }
+}
+
+import tree._
+
+package infra {
+  class Inject[K, V](val keyOrdering: Ordering[K], val valueMonoid: Monoid[V])
+    extends Serializable {
+    def iNode(clr: Color, dat: Data[K], ls: Node[K], rs: Node[K]) =
+      new Inject[K, V](keyOrdering, valueMonoid) with INodeInc[K, V] with IncrementMap[K, V] {
+        // INode[K]
+        val color = clr
+        val lsub = ls.asInstanceOf[NodeInc[K, V]]
+        val rsub = rs.asInstanceOf[NodeInc[K, V]]
+        val data = dat.asInstanceOf[DataMap[K, V]]
+      }
+  }
+}
+
+import infra._
+
+/**
+ * An inheritable and mixable trait for adding increment operation to ordered maps
+ * @tparam K The key type
+ * @tparam V The value type
+ * @tparam IN The node type of the concrete internal R/B tree subclass
+ * @tparam M The map self-type of the concrete map subclass
+ */
+trait IncrementMapLike[K, V, IN <: INodeInc[K, V], M <: IncrementMapLike[K, V, IN, M] with SortedMap[K, V]]
+  extends NodeInc[K, V] with OrderedMapLike[K, V, IN, M] {
+
+  /**
+   * Add (w.r.t. valueMonoid) a given value to the value currently stored at key.
+   * @note If key is not present, equivalent to insert(k, valueMonoid.plus(valueMonoid.zero, iv)
+   */
+  def increment(k: K, iv: V) = this.incr(
+    new DataMap[K, V] {
+      val key = k
+      val value = iv
+    }).asInstanceOf[M]
```

Comment:
It is being used so that the `xxxMap` classes can return their own type (encoded by self type `M`).   As far as I know, the only way to avoid it would be to extend the self types up into the tree classes.  I made a few attempts at that, and could not get any of them to work, and so in the end I decided that casting in a few places was acceptable as a way to get it all working.

The cast succeeds because the class-specific implementation of `iNode` generates objects of the correct type.

Generally speaking, I share your aesthetic of wishing to avoid `asInstanceOf`, I just wasn't able to figure out how to avoid it in the places I did use it.  The places I do use it, the type casts will succeed.

That said, if I or anybody else can demonstrate a version of the code that doesn't need them, that would be very nice.